File: scripts/install_root.py
#!/usr/bin/env -S uv run --script
# /// script
# requires-python = ">=3.11"
# dependencies = []
# ///
"""Read-only Agentera install-root classification.

This module owns install-root identity and diagnostics without mutating durable
bundle state. Callers may use the structured result directly or render the
diagnostic message for humans with ``format_diagnostic``.
"""

from __future__ import annotations

import json
import os
import sys
from dataclasses import asdict, dataclass
from pathlib import Path
from typing import Any, Mapping

BUNDLE_MARKER = ".agentera-bundle.json"
SETUP_EVIDENCE = (
    "scripts/validate_capability.py",
    "hooks",
    "skills",
    "skills/agentera/SKILL.md",
)
BUNDLE_EVIDENCE = (
    "scripts/agentera",
    "skills/agentera/SKILL.md",
    "registry.json",
    BUNDLE_MARKER,
)
MANAGED_EVIDENCE = tuple(dict.fromkeys((*SETUP_EVIDENCE, *BUNDLE_EVIDENCE)))
SOURCE_LABELS = {
    "explicit": "explicit --install-root",
    "environment": "AGENTERA_HOME",
    "default": "default app home",
}


@dataclass(frozen=True)
class Diagnostic:
    code: str
    severity: str
    message: str
    evidence: dict[str, Any]


@dataclass(frozen=True)
class Classification:
    source: str
    source_label: str
    path: str
    kind: str
    safe_action: str
    diagnostic: Diagnostic
    managed_status: str
    stale_status: str
    missing_evidence: list[str]
    expected_version: str | None = None
    current_version: str | None = None

    def to_dict(self) -> dict[str, Any]:
        return asdict(self)


def resolve_candidate(
    explicit_root: str | Path | None,
    *,
    env: Mapping[str, str],
    home: Path,
) -> tuple[Path, str]:
    """Resolve the candidate path using explicit, environment, then default precedence."""
    if explicit_root is not None:
        return Path(explicit_root).expanduser().resolve(), "explicit"
    configured = env.get("AGENTERA_HOME")
    if configured:
        return Path(configured).expanduser().resolve(), "environment"
    default = env.get("AGENTERA_DEFAULT_INSTALL_ROOT")
    if default:
        return Path(default).expanduser().resolve(), "default"
    return _default_app_home(env, home).resolve(), "default"


def _default_app_home(env: Mapping[str, str], home: Path) -> Path:
    """Return the platform data-home default for the Agentera app home."""
    if sys.platform == "darwin":
        return home.expanduser() / "Library" / "Application Support" / "agentera"
    if os.name == "nt":
        appdata = env.get("APPDATA")
        base = Path(appdata).expanduser() if appdata else home.expanduser() / "AppData" / "Roaming"
        return base / "agentera"
    xdg = env.get("XDG_DATA_HOME")
    base = Path(xdg).expanduser() if xdg else home.expanduser() / ".local" / "share"
    return base / "agentera"


def _linux_default_app_home(env: Mapping[str, str], home: Path) -> Path:
    xdg = env.get("XDG_DATA_HOME")
    base = Path(xdg).expanduser() if xdg else home.expanduser() / ".local" / "share"
    return base / "agentera"


def _macos_default_app_home(home: Path) -> Path:
    return home.expanduser() / "Library" / "Application Support" / "agentera"


def _windows_default_app_home(env: Mapping[str, str], home: Path) -> Path:
    appdata = env.get("APPDATA")
    base = Path(appdata).expanduser() if appdata else home.expanduser() / "AppData" / "Roaming"
    return base / "agentera"


def known_platform_default_app_homes(env: Mapping[str, str], home: Path) -> frozenset[Path]:
    """Return every OS-native default app-home path, regardless of the current platform."""
    return frozenset(
        {
            _linux_default_app_home(env, home).expanduser().resolve(),
            _macos_default_app_home(home).expanduser().resolve(),
            _windows_default_app_home(env, home).expanduser().resolve(),
        }
    )


def is_foreign_platform_default_app_home(
    candidate: Path,
    *,
    env: Mapping[str, str],
    home: Path,
) -> bool:
    """Return true when ``candidate`` is another platform's default app home."""
    resolved = candidate.expanduser().resolve()
    platform_default = _default_app_home(env, home).expanduser().resolve()
    if resolved == platform_default:
        return False
    return resolved in known_platform_default_app_homes(env, home)


def classify_resolved_root(
    root: Path,
    *,
    source: str,
    expected_version: str | None = None,
) -> Classification:
    """Classify a resolved root path without creating, deleting, or modifying files."""
    root = root.expanduser().resolve()
    source_label = SOURCE_LABELS.get(source, source)
    expected = expected_version

    if not root.exists():
        if source == "default":
            return _classification(
                root,
                source,
                "missing_default",
                "preview_refresh",
                "missing",
                "stale",
                ["directory", "managed bundle evidence"],
                Diagnostic(
                    "install_root.missing_default_root",
                    "warning",
                    "Agentera is not installed in the normal directory yet; a preview can show the repair before anything changes",
                    {"path": str(root), "source": source_label},
                ),
                expected_version=expected,
            )
        return _classification(
            root,
            source,
            "missing_explicit_or_environment",
            "require_existing_managed_root",
            "missing",
            "not_applicable",
            ["directory", "managed bundle evidence"],
            Diagnostic(
                "install_root.missing_selected_root",
                "error",
                "Agentera was told to use a directory that does not exist",
                {"path": str(root), "source": source_label},
            ),
            expected_version=expected,
        )

    if root.is_file():
        return _classification(
            root,
            source,
            "file_valued_root",
            "reject_file_path",
            "invalid",
            "not_applicable",
            ["directory"],
            Diagnostic(
                "install_root.file_path",
                "error",
                "Agentera was told to use a file instead of a directory",
                {"path": str(root), "source": source_label},
            ),
            expected_version=expected,
        )

    setup_missing = _missing_entries(root, SETUP_EVIDENCE)
    bundle_missing = _missing_entries(root, BUNDLE_EVIDENCE)
    has_setup_evidence = not setup_missing
    has_bundle_evidence = not bundle_missing
    present_evidence = [entry for entry in MANAGED_EVIDENCE if (root / entry).exists()]
    marker = _read_bundle_marker(root)
    current = marker.get("version") if marker else None

    if has_setup_evidence and not (root / BUNDLE_MARKER).exists():
        return _managed_fresh(root, source, setup_missing, bundle_missing, expected, current)

    if has_bundle_evidence:
        if expected is not None and current != expected:
            return _managed_stale(
                root,
                source,
                expected,
                current,
                missing_evidence=["current bundle marker/version or required CLI command evidence"],
                reason="version_mismatch",
                evidence={"expectedVersion": expected, "currentVersion": current, "markerPath": str(root / BUNDLE_MARKER)},
            )
        command_missing = _missing_script_commands(root / "scripts" / "agentera", ("hej",))
        if command_missing:
            return _managed_stale(
                root,
                source,
                expected,
                current,
                missing_evidence=["current bundle marker/version or required CLI command evidence"],
                reason="missing_command",
                evidence={"missingCommands": command_missing, "scriptPath": str(root / "scripts" / "agentera")},
            )
        return _managed_fresh(root, source, setup_missing, bundle_missing, expected, current)

    managed_without_marker = (
        (root / "scripts" / "agentera").exists()
        and (root / "skills" / "agentera" / "SKILL.md").exists()
        and (root / "registry.json").exists()
    )
    if managed_without_marker:
        return _managed_stale(
            root,
            source,
            expected,
            current,
            missing_evidence=["current bundle marker/version or required CLI command evidence"],
            reason="missing_marker",
            evidence={"expectedVersion": expected, "currentVersion": current, "markerPath": str(root / BUNDLE_MARKER)},
        )

    if present_evidence:
        return _classification(
            root,
            source,
            "invalid_bundle",
            "reject_invalid_bundle",
            "invalid",
            "unknown",
            ["valid bundle marker and complete managed bundle evidence"],
            Diagnostic(
                "install_root.invalid_bundle",
                "error",
                "This directory looks like a broken Agentera install",
                {
                    "path": str(root),
                    "source": source_label,
                    "presentEvidence": present_evidence,
                    "missingSetupEvidence": setup_missing,
                    "missingBundleEvidence": bundle_missing,
                },
            ),
            expected_version=expected,
            current_version=current,
        )

    return _classification(
        root,
        source,
        "unmanaged_directory",
        "reject_unmanaged_directory",
        "unmanaged",
        "not_applicable",
        ["managed bundle evidence"],
        Diagnostic(
            "install_root.unmanaged_directory",
            "error",
            "This directory already has files Agentera does not recognize",
            {"path": str(root), "source": source_label},
        ),
        expected_version=expected,
    )


def classify_install_root(
    explicit_root: str | Path | None,
    *,
    env: Mapping[str, str],
    home: Path,
    expected_version: str | None = None,
) -> Classification:
    root, source = resolve_candidate(explicit_root, env=env, home=home)
    return classify_resolved_root(root, source=source, expected_version=expected_version)


def format_diagnostic(classification: Classification) -> str:
    """Return human display text without collapsing the structured diagnostic."""
    return f"{classification.diagnostic.severity}: {classification.diagnostic.message} ({classification.path})"


def _classification(
    root: Path,
    source: str,
    kind: str,
    safe_action: str,
    managed_status: str,
    stale_status: str,
    missing_evidence: list[str],
    diagnostic: Diagnostic,
    *,
    expected_version: str | None = None,
    current_version: str | None = None,
) -> Classification:
    return Classification(
        source=source,
        source_label=SOURCE_LABELS.get(source, source),
        path=str(root),
        kind=kind,
        safe_action=safe_action,
        diagnostic=diagnostic,
        managed_status=managed_status,
        stale_status=stale_status,
        missing_evidence=missing_evidence,
        expected_version=expected_version,
        current_version=current_version,
    )


def _managed_fresh(
    root: Path,
    source: str,
    setup_missing: list[str],
    bundle_missing: list[str],
    expected: str | None,
    current: str | None,
) -> Classification:
    return _classification(
        root,
        source,
        "managed_fresh",
        "use_root",
        "managed",
        "fresh",
        [],
        Diagnostic(
            "install_root.managed_fresh",
            "info",
            "Agentera app files are ready",
            {
                "path": str(root),
                "source": SOURCE_LABELS.get(source, source),
                "expectedVersion": expected,
                "currentVersion": current,
                "missingSetupEvidence": setup_missing,
                "missingBundleEvidence": bundle_missing,
            },
        ),
        expected_version=expected,
        current_version=current,
    )


def _managed_stale(
    root: Path,
    source: str,
    expected: str | None,
    current: str | None,
    *,
    missing_evidence: list[str],
    reason: str,
    evidence: dict[str, Any],
) -> Classification:
    return _classification(
        root,
        source,
        "managed_stale",
        "preview_refresh",
        "managed",
        "stale",
        missing_evidence,
        Diagnostic(
            "install_root.managed_stale",
            "warning",
            "Agentera app files need repair",
            {"path": str(root), "source": SOURCE_LABELS.get(source, source), "reason": reason, **evidence},
        ),
        expected_version=expected,
        current_version=current,
    )


def _missing_entries(root: Path, entries: tuple[str, ...]) -> list[str]:
    return [entry for entry in entries if not (root / entry).exists()]


def _read_bundle_marker(root: Path) -> dict[str, Any] | None:
    marker = root / BUNDLE_MARKER
    if not marker.is_file():
        return None
    try:
        data = json.loads(marker.read_text(encoding="utf-8"))
    except (OSError, json.JSONDecodeError):
        return None
    return data if isinstance(data, dict) else None


def _missing_script_commands(script: Path, commands: tuple[str, ...]) -> list[str]:
    if not script.is_file():
        return list(commands)
    try:
        text = script.read_text(encoding="utf-8")
    except OSError:
        return list(commands)
    return [command for command in commands if command not in text]


def format_resolved_app_home(
    explicit_root: str | Path | None = None,
    *,
    env: Mapping[str, str] | None = None,
    home: Path | None = None,
    output_format: str = "text",
) -> str:
    """Return the resolved app home for agent/bootstrap callers."""
    resolved_env = env if env is not None else os.environ
    resolved_home = home if home is not None else Path.home()
    path, source = resolve_candidate(explicit_root, env=resolved_env, home=resolved_home)
    if output_format == "json":
        return json.dumps(
            {
                "path": str(path),
                "source": source,
                "sourceLabel": SOURCE_LABELS.get(source, source),
                "platform": sys.platform,
            }
        )
    return str(path)


def main(argv: list[str] | None = None) -> int:
    import argparse

    parser = argparse.ArgumentParser(description="Resolve Agentera app-home paths.")
    parser.add_argument(
        "--format",
        choices=["text", "json"],
        default="text",
        help="Output format (default: text prints the resolved app-home path only)",
    )
    parser.add_argument(
        "--install-root",
        type=Path,
        default=None,
        help="Explicit app home to resolve (default: AGENTERA_HOME, then platform data home)",
    )
    parser.add_argument(
        "--home",
        type=Path,
        default=Path.home(),
        help="Home directory used for the default app home",
    )
    args = parser.parse_args(argv)
    print(
        format_resolved_app_home(
            args.install_root,
            env=os.environ,
            home=args.home,
            output_format=args.format,
        )
    )
    return 0


if __name__ == "__main__":
    raise SystemExit(main())

File: packages/cli/src/state/installRoot.ts
import fs from "node:fs";
import path from "node:path";

import { expanduser, isFile, pathExists, resolvePath } from "../core/paths.js";

/**
 * Read-only Agentera install-root classification. Faithful TypeScript port of
 * `scripts/install_root.py`. Owns install-root identity and diagnostics without
 * mutating durable bundle state.
 */

export const BUNDLE_MARKER = ".agentera-bundle.json";

export const SETUP_EVIDENCE = [
  "skills",
  "skills/agentera/SKILL.md",
  "registry.json",
] as const;

export const BUNDLE_EVIDENCE = [
  "skills/agentera/SKILL.md",
  "registry.json",
  BUNDLE_MARKER,
] as const;

// Python: tuple(dict.fromkeys((*SETUP_EVIDENCE, *BUNDLE_EVIDENCE))) — order-preserving dedupe.
export const MANAGED_EVIDENCE: readonly string[] = ((): string[] => {
  const seen = new Set<string>();
  const out: string[] = [];
  for (const entry of [...SETUP_EVIDENCE, ...BUNDLE_EVIDENCE]) {
    if (!seen.has(entry)) {
      seen.add(entry);
      out.push(entry);
    }
  }
  return out;
})();

export const SOURCE_LABELS: Record<string, string> = {
  explicit: "explicit --install-root",
  environment: "AGENTERA_HOME",
  default: "default app home",
};

export interface Diagnostic {
  code: string;
  severity: string;
  message: string;
  evidence: Record<string, unknown>;
}

export interface Classification {
  source: string;
  source_label: string;
  path: string;
  kind: string;
  safe_action: string;
  diagnostic: Diagnostic;
  managed_status: string;
  stale_status: string;
  missing_evidence: string[];
  expected_version: string | null;
  current_version: string | null;
}

/** Mirror of `Classification.to_dict()` (asdict): plain object, already shaped. */
export function toDict(c: Classification): Record<string, unknown> {
  return {
    source: c.source,
    source_label: c.source_label,
    path: c.path,
    kind: c.kind,
    safe_action: c.safe_action,
    diagnostic: {
      code: c.diagnostic.code,
      severity: c.diagnostic.severity,
      message: c.diagnostic.message,
      evidence: c.diagnostic.evidence,
    },
    managed_status: c.managed_status,
    stale_status: c.stale_status,
    missing_evidence: c.missing_evidence,
    expected_version: c.expected_version,
    current_version: c.current_version,
  };
}

export interface PlatformEnv {
  /** node-style platform string; defaults to process.platform. */
  platform?: NodeJS.Platform;
}

function sourceLabel(source: string): string {
  return SOURCE_LABELS[source] ?? source;
}

export function resolveCandidate(
  explicitRoot: string | null,
  opts: { env: Record<string, string | undefined>; home: string } & PlatformEnv,
): [string, string] {
  const { env, home } = opts;
  if (explicitRoot !== null && explicitRoot !== undefined) {
    return [resolvePath(explicitRoot), "explicit"];
  }
  const configured = env.AGENTERA_HOME;
  if (configured) {
    return [resolvePath(configured), "environment"];
  }
  const def = env.AGENTERA_DEFAULT_INSTALL_ROOT;
  if (def) {
    return [resolvePath(def), "default"];
  }
  return [resolvePath(defaultAppHome(env, home, opts.platform)), "default"];
}

export function defaultAppHome(
  env: Record<string, string | undefined>,
  home: string,
  platform: NodeJS.Platform = process.platform,
): string {
  if (platform === "darwin") {
    return macosDefaultAppHome(home);
  }
  if (platform === "win32") {
    return windowsDefaultAppHome(env, home);
  }
  return linuxDefaultAppHome(env, home);
}

function linuxDefaultAppHome(env: Record<string, string | undefined>, home: string): string {
  const xdg = env.XDG_DATA_HOME;
  const base = xdg ? expanduser(xdg) : path.join(expanduser(home), ".local", "share");
  return path.join(base, "agentera");
}

function macosDefaultAppHome(home: string): string {
  return path.join(expanduser(home), "Library", "Application Support", "agentera");
}

function windowsDefaultAppHome(env: Record<string, string | undefined>, home: string): string {
  const appdata = env.APPDATA;
  const base = appdata
    ? expanduser(appdata)
    : path.join(expanduser(home), "AppData", "Roaming");
  return path.join(base, "agentera");
}

export function knownPlatformDefaultAppHomes(
  env: Record<string, string | undefined>,
  home: string,
): Set<string> {
  return new Set<string>([
    resolvePath(linuxDefaultAppHome(env, home)),
    resolvePath(macosDefaultAppHome(home)),
    resolvePath(windowsDefaultAppHome(env, home)),
  ]);
}

export function isForeignPlatformDefaultAppHome(
  candidate: string,
  opts: { env: Record<string, string | undefined>; home: string } & PlatformEnv,
): boolean {
  const resolved = resolvePath(candidate);
  const platformDefault = resolvePath(defaultAppHome(opts.env, opts.home, opts.platform));
  if (resolved === platformDefault) {
    return false;
  }
  return knownPlatformDefaultAppHomes(opts.env, opts.home).has(resolved);
}

function classification(
  root: string,
  source: string,
  kind: string,
  safeAction: string,
  managedStatus: string,
  staleStatus: string,
  missingEvidence: string[],
  diagnostic: Diagnostic,
  opts: { expectedVersion?: string | null; currentVersion?: string | null } = {},
): Classification {
  return {
    source,
    source_label: sourceLabel(source),
    path: root,
    kind,
    safe_action: safeAction,
    diagnostic,
    managed_status: managedStatus,
    stale_status: staleStatus,
    missing_evidence: missingEvidence,
    expected_version: opts.expectedVersion ?? null,
    current_version: opts.currentVersion ?? null,
  };
}

function managedFresh(
  root: string,
  source: string,
  setupMissing: string[],
  bundleMissing: string[],
  expected: string | null,
  current: string | null,
): Classification {
  return classification(
    root,
    source,
    "managed_fresh",
    "use_root",
    "managed",
    "fresh",
    [],
    {
      code: "install_root.managed_fresh",
      severity: "info",
      message: "Agentera app files are ready",
      evidence: {
        path: root,
        source: sourceLabel(source),
        expectedVersion: expected,
        currentVersion: current,
        missingSetupEvidence: setupMissing,
        missingBundleEvidence: bundleMissing,
      },
    },
    { expectedVersion: expected, currentVersion: current },
  );
}

function managedStaleDiagnosticMessage(reason: string): string {
  if (reason === "version_mismatch") {
    return "Agentera app files are valid but need an update to the expected version";
  }
  return "Agentera app files need repair";
}

function managedStale(
  root: string,
  source: string,
  expected: string | null,
  current: string | null,
  missingEvidence: string[],
  reason: string,
  evidence: Record<string, unknown>,
): Classification {
  return classification(
    root,
    source,
    "managed_stale",
    "preview_refresh",
    "managed",
    "stale",
    missingEvidence,
    {
      code: "install_root.managed_stale",
      severity: "warning",
      message: managedStaleDiagnosticMessage(reason),
      evidence: { path: root, source: sourceLabel(source), reason, ...evidence },
    },
    { expectedVersion: expected, currentVersion: current },
  );
}

function missingEntries(root: string, entries: readonly string[]): string[] {
  return entries.filter((entry) => !pathExists(path.join(root, entry)));
}

function readBundleMarker(root: string): Record<string, unknown> | null {
  const marker = path.join(root, BUNDLE_MARKER);
  if (!isFile(marker)) {
    return null;
  }
  try {
    const data = JSON.parse(fs.readFileSync(marker, "utf8"));
    return data && typeof data === "object" && !Array.isArray(data)
      ? (data as Record<string, unknown>)
      : null;
  } catch {
    return null;
  }
}

function missingScriptCommands(script: string, commands: readonly string[]): string[] {
  if (!isFile(script)) {
    return [...commands];
  }
  let text: string;
  try {
    text = fs.readFileSync(script, "utf8");
  } catch {
    return [...commands];
  }
  return commands.filter((command) => !text.includes(command));
}

export function classifyResolvedRoot(
  rootInput: string,
  opts: { source: string; expectedVersion?: string | null },
): Classification {
  const root = resolvePath(rootInput);
  const source = opts.source;
  const label = sourceLabel(source);
  const expected = opts.expectedVersion ?? null;

  if (!pathExists(root)) {
    if (source === "default") {
      return classification(
        root,
        source,
        "missing_default",
        "preview_refresh",
        "missing",
        "stale",
        ["directory", "managed bundle evidence"],
        {
          code: "install_root.missing_default_root",
          severity: "warning",
          message:
            "Agentera is not installed in the normal directory yet; a preview can show the repair before anything changes",
          evidence: { path: root, source: label },
        },
        { expectedVersion: expected },
      );
    }
    return classification(
      root,
      source,
      "missing_explicit_or_environment",
      "require_existing_managed_root",
      "missing",
      "not_applicable",
      ["directory", "managed bundle evidence"],
      {
        code: "install_root.missing_selected_root",
        severity: "error",
        message: "Agentera was told to use a directory that does not exist",
        evidence: { path: root, source: label },
      },
      { expectedVersion: expected },
    );
  }

  if (isFile(root)) {
    return classification(
      root,
      source,
      "file_valued_root",
      "reject_file_path",
      "invalid",
      "not_applicable",
      ["directory"],
      {
        code: "install_root.file_path",
        severity: "error",
        message: "Agentera was told to use a file instead of a directory",
        evidence: { path: root, source: label },
      },
      { expectedVersion: expected },
    );
  }

  const setupMissing = missingEntries(root, SETUP_EVIDENCE);
  const bundleMissing = missingEntries(root, BUNDLE_EVIDENCE);
  const hasSetupEvidence = setupMissing.length === 0;
  const hasBundleEvidence = bundleMissing.length === 0;
  const presentEvidence = MANAGED_EVIDENCE.filter((entry) => pathExists(path.join(root, entry)));
  const marker = readBundleMarker(root);
  const current = (marker?.version as string | undefined) ?? null;

  if (hasSetupEvidence && !pathExists(path.join(root, BUNDLE_MARKER))) {
    return managedFresh(root, source, setupMissing, bundleMissing, expected, current);
  }

  if (hasBundleEvidence) {
    if (expected !== null && current !== expected) {
      return managedStale(
        root,
        source,
        expected,
        current,
        ["current bundle marker/version or required CLI command evidence"],
        "version_mismatch",
        {
          expectedVersion: expected,
          currentVersion: current,
          markerPath: path.join(root, BUNDLE_MARKER),
        },
      );
    }
    // Node self-contained model: a marked bundle with current data is fresh; the
    // Python CLI command-probe (scripts/agentera) no longer applies.
    return managedFresh(root, source, setupMissing, bundleMissing, expected, current);
  }

  const managedWithoutMarker =
    pathExists(path.join(root, "scripts", "agentera")) &&
    pathExists(path.join(root, "skills", "agentera", "SKILL.md")) &&
    pathExists(path.join(root, "registry.json"));
  if (managedWithoutMarker) {
    return managedStale(
      root,
      source,
      expected,
      current,
      ["current bundle marker/version or required CLI command evidence"],
      "missing_marker",
      {
        expectedVersion: expected,
        currentVersion: current,
        markerPath: path.join(root, BUNDLE_MARKER),
      },
    );
  }

  if (presentEvidence.length > 0) {
    return classification(
      root,
      source,
      "invalid_bundle",
      "reject_invalid_bundle",
      "invalid",
      "unknown",
      ["valid bundle marker and complete managed bundle evidence"],
      {
        code: "install_root.invalid_bundle",
        severity: "error",
        message: "This directory looks like a broken Agentera install",
        evidence: {
          path: root,
          source: label,
          presentEvidence,
          missingSetupEvidence: setupMissing,
          missingBundleEvidence: bundleMissing,
        },
      },
      { expectedVersion: expected, currentVersion: current },
    );
  }

  return classification(
    root,
    source,
    "unmanaged_directory",
    "reject_unmanaged_directory",
    "unmanaged",
    "not_applicable",
    ["managed bundle evidence"],
    {
      code: "install_root.unmanaged_directory",
      severity: "error",
      message: "This directory already has files Agentera does not recognize",
      evidence: { path: root, source: label },
    },
    { expectedVersion: expected },
  );
}

export function classifyInstallRoot(
  explicitRoot: string | null,
  opts: {
    env: Record<string, string | undefined>;
    home: string;
    expectedVersion?: string | null;
  } & PlatformEnv,
): Classification {
  const [root, source] = resolveCandidate(explicitRoot, opts);
  return classifyResolvedRoot(root, { source, expectedVersion: opts.expectedVersion ?? null });
}

export function formatDiagnostic(c: Classification): string {
  return `${c.diagnostic.severity}: ${c.diagnostic.message} (${c.path})`;
}

/** Resolved app-home path for agent/bootstrap callers (port of install_root.py). */
export function formatResolvedAppHome(
  explicitRoot: string | null | undefined,
  opts: {
    env?: Record<string, string | undefined>;
    home?: string;
    format?: "text" | "json";
    platform?: NodeJS.Platform;
  } = {},
): string {
  const env = opts.env ?? process.env;
  const home = opts.home ?? expanduser("~");
  const platform = opts.platform ?? process.platform;
  const [resolvedPath, source] = resolveCandidate(explicitRoot ?? null, { env, home, platform });
  if ((opts.format ?? "text") === "json") {
    return JSON.stringify({
      path: resolvedPath,
      source,
      sourceLabel: sourceLabel(source),
      platform,
    });
  }
  return resolvedPath;
}

